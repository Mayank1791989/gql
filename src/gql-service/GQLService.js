/* @flow */
import GQLSchemaService from 'gql-schema-service';
import GQLQueryService from 'gql-query-service';
import GQLConfig from 'gql-config';
import GQLWatcher from 'gql-watcher';
import GQLBaseService from 'gql-shared/GQLBaseService';
import GQLSchema from 'gql-shared/GQLSchema';
import log from 'gql-shared/log';

const logger = log.getLogger('gql');

import {
  type GQLHint,
  type GQLInfo,
  type GQLLocation,
  type GQLPosition,
} from 'gql-shared/types';

import { type GQLError } from 'gql-shared/GQLError';
import getResolverLocation from 'gql-schema-service/commands/getResolverLocation';
import getSchemaLocation from 'gql-schema-service/commands/getSchemaLocation';

type Options = $ReadOnly<{|
  configDir?: string,

  // watch
  watchman?: boolean,
  watch?: boolean,
|}>;

type CommandParams = $ReadOnly<{|
  sourcePath: string,
  sourceText: string,
  position: GQLPosition,
|}>;

type GetResolverOrSchemaParams = $ReadOnly<{|
  fieldType: string,
  fieldName: string,
|}>;

export type { Options as GQLServiceOptions };

export function isGql(sourcePath: string) {
  return !(sourcePath.endsWith('ts') || sourcePath.endsWith('ts'));
}

export default class GQLService extends GQLBaseService {
  _schemaService: GQLSchemaService;
  _queryService: ?GQLQueryService;

  _config: GQLConfig;
  _watcher: GQLWatcher;

  constructor(_options: ?Options) {
    super();
    const options = _options || {};

    this._config = new GQLConfig({
      configDir: options.configDir || process.cwd(),
    });
    this._watcher = new GQLWatcher({
      watchman: options.watchman,
      watch: options.watch,
    });

    // setup schema service
    this._schemaService = new GQLSchemaService({
      config: this._config,
      watcher: this._watcher,
    });
    this._schemaService.onChange(this._triggerChange);
    this._schemaService.onError(this._triggerError);

    // setup query service
    if (this._config.getQueryConfig()) {
      const queryService = new GQLQueryService({
        config: this._config,
        schemaService: this._schemaService,
        watcher: this._watcher,
      });
      queryService.onChange(this._triggerChange);
      queryService.onError(this._triggerError);
      this._queryService = queryService;
    }
  }

  onLog(listener) {
    return log.onLog(listener);
  }

  async _handleStart() {
    // start schema service
    await this._schemaService.start();
    if (this._queryService) {
      await this._queryService.start();
    }
  }

  async _handleStop() {
    await this._schemaService.stop();
    if (this._queryService) {
      await this._queryService.stop();
    }
  }

  getSchema(): GQLSchema {
    return this._schemaService.getSchema();
  }

  getConfig(): GQLConfig {
    return this._config;
  }

  status(): Array<GQLError> {
    if (!this._isRunning) {
      return [];
    }
    try {
      const schemaErrors = this._schemaService.getSchemaErrors();
      const queryErrors = this._queryService
        ? this._queryService.getErrors()
        : [];
      return schemaErrors.concat(queryErrors.filter(err => Boolean(err)));
    } catch (err) {
      this._triggerError(err);
      return [];
    }
  }

  autocomplete(params: CommandParams): Array<GQLHint> {
    return this._catchThrownErrors(() => {
      logger.debug('autocomplete request');
      logger.time('autocomplete response');
      if (!this._isRunning) {
        return [];
      }

      // codemirror instance
      let results = [];
      logger.time('getFileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      logger.timeEnd('getFileConfig');
      logger.debug('FileType detected:', fileConfig && fileConfig.type);

      logger.time('getHintsAtPosition');
      if (fileConfig && fileConfig.type === 'schema') {
        results = this._schemaService.getHintsAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      if (this._queryService && fileConfig && fileConfig.type === 'query') {
        results = this._queryService.getHintsAtPosition({
          fileContent: params.sourceText,
          filePath: params.sourcePath,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      logger.timeEnd('getHintsAtPosition');
      logger.timeEnd('autocomplete response');
      return results;
    }, []);
  }

  getLenses(params: CommandParams): ?GQLLocation {
    return this._catchThrownErrors(async () => {
      logger.debug('getLenses request');
      logger.time('getLenses response');
      if (!this._isRunning) {
        return null;
      }

      let defLenses = null;

      if (isGql(params.sourcePath)) {
        logger.time('getFileConfig');
        const fileConfig = this._config.getFileConfig(params.sourcePath);
        logger.timeEnd('getFileConfig');
        logger.debug('FileType detected:', fileConfig && fileConfig.type);

        logger.time('getLenses');

        if (fileConfig && fileConfig.type === 'schema') {
          defLenses = await this._schemaService.getLenses({
            filePath: params.sourcePath,
          });
        }
      } else {
        logger.time('getLenses');
        defLenses = await this._schemaService.getLenses({
          filePath: params.sourcePath,
        });
      }

      logger.timeEnd('getLenses');
      logger.timeEnd('getLenses response');
      return defLenses;
    }, null);
  }

  getResolverLocation(params: GetResolverOrSchemaParams): ?GQLLocation {
    return this._catchThrownErrors(async () => {
      logger.debug('getResolverLocation request');
      logger.time('getResolverLocation response');
      if (!this._isRunning) {
        return null;
      }

      logger.time('getResolverLocation');

      const schema = this._schemaService.getSchema();
      const resolverLocation = await getResolverLocation(
        params.fieldType,
        params.fieldName,
        schema,
        this._config.getResolverConfig().files,
        this._config.getResolverConfig().baseDir,
      );

      logger.timeEnd('getResolverLocation');
      logger.timeEnd('getResolverLocation response');
      return resolverLocation;
    }, null);
  }

  getSchemaLocation(params: GetResolverOrSchemaParams): ?GQLLocation {
    return this._catchThrownErrors(() => {
      logger.debug('getSchemaLocation request');
      logger.time('getSchemaLocation response');
      if (!this._isRunning) {
        return null;
      }

      logger.time('getSchemaLocation');

      const schema = this._schemaService.getSchema();
      const schemaLocation = getSchemaLocation(
        params.fieldType,
        params.fieldName,
        schema,
      );

      logger.timeEnd('getSchemaLocation');
      logger.timeEnd('getSchemaLocation response');
      return schemaLocation;
    }, null);
  }

  getDef(params: CommandParams): ?GQLLocation {
    return this._catchThrownErrors(async () => {
      logger.debug('getDef request');
      logger.time('getDef response');
      if (!this._isRunning) {
        return null;
      }

      let defLocation = null;

      if (isGql(params.sourcePath)) {
        logger.time('getFileConfig');
        const fileConfig = this._config.getFileConfig(params.sourcePath);
        logger.timeEnd('getFileConfig');
        logger.debug('FileType detected:', fileConfig && fileConfig.type);

        logger.time('getDef');
        if (fileConfig && fileConfig.type === 'schema') {
          defLocation = await this._schemaService.getDefinitionAtPosition({
            fileContent: params.sourceText,
            filePath: params.sourcePath,
            fileOptions: fileConfig.opts,
            position: params.position,
            resolversGlob: this._config.getResolverConfig().files,
            resolversBaseDir: this._config.getResolverConfig().baseDir,
          });
        }

        if (this._queryService && fileConfig && fileConfig.type === 'query') {
          defLocation = await this._queryService.getDefinitionAtPosition({
            fileContent: params.sourceText,
            filePath: params.sourcePath,
            fileOptions: fileConfig.opts,
            position: params.position,
          });
        }
      } else {
        logger.time('getDef');
        defLocation = await this._schemaService.getDefinitionAtPosition({
          fileContent: params.sourceText,
          filePath: params.sourcePath,
          position: params.position,
        });
      }
      logger.timeEnd('getDef');
      logger.timeEnd('getDef response');
      return defLocation;
    }, null);
  }

  findRefs(params: CommandParams): Array<GQLLocation> {
    return this._catchThrownErrors(() => {
      logger.debug('findRefs request');
      logger.time('findRefs response');
      if (!this._isRunning) {
        return [];
      }

      let refLocations = [];
      logger.time('getFileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      logger.timeEnd('getFileConfig');
      logger.debug('FileType detected:', fileConfig && fileConfig.type);

      logger.time('findRefs');
      if (fileConfig && fileConfig.type === 'schema') {
        refLocations = this._schemaService.findRefsOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      // if (this._queryService && fileConfig && fileConfig.type === 'query') {
      //   refLocations = this._queryService.findRefsOfTokenAtPosition({
      //     fileContent: params.sourceText,
      //     filePath: params.sourcePath,
      //     fileOptions: fileConfig.opts,
      //     position: params.position,
      //   });
      // }
      logger.timeEnd('findRefs');
      logger.timeEnd('findRefs response');
      // @TODO query not implemented
      return refLocations;
    }, []);
  }

  getInfo(params: CommandParams): ?GQLInfo {
    return this._catchThrownErrors(() => {
      logger.debug('getInfo request');
      logger.time('getInfo response');
      if (!this._isRunning) {
        return null;
      }

      let info = null;
      logger.time('getFileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      logger.timeEnd('getFileConfig');
      logger.debug('FileType detected:', fileConfig && fileConfig.type);

      logger.time('getInfoAtToken');
      if (fileConfig && fileConfig.type === 'schema') {
        info = this._schemaService.getInfoOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }

      if (this._queryService && fileConfig && fileConfig.type === 'query') {
        info = this._queryService.getInfoOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          filePath: params.sourcePath,
          position: params.position,
        });
      }
      logger.timeEnd('getInfoAtToken');
      logger.timeEnd('getInfo response');
      return info;
    }, null);
  }
}
