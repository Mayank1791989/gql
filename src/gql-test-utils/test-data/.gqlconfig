{
  schema: {
    files: "schema/**/*.gql",
    graphQLOptions: {
      commentDescriptions: true,
    },
  }
}
