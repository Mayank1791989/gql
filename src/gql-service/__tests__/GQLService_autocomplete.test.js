/* @flow */
import code from 'gql-test-utils/code';
import GQLService from '../GQLService';
import { getSchemaFiles, getHints, sortHints } from 'gql-test-utils/test-data';
import { createTempFiles } from 'gql-test-utils/file';
import path from 'path';

describe('Schema: autocomplete', () => {
  it('works in schema files', async () => {
    const rootPath = createTempFiles({
      '.gqlconfig': `
          {
            schema: {
              files: 'schema/*.gql',
            }
          }
        `,
      'schema/schema.gql': `
          type Query {
            viewer: Viewer
          }

          type Viewer {
            name: String
          }
        `,
    });

    const gql = new GQLService({
      configDir: rootPath,
      watch: false,
    });
    gql.onError(err => {
      throw err;
    });

    await gql.start();

    expect(
      gql
        .autocomplete({
          sourcePath: path.join(rootPath, 'schema/user.gql'),
          ...code(`
            type User {
              viewer: Vi
              #---------^
            }
        `),
        })
        .sort(sortHints),
    ).toMatchSnapshot();
  });

  it('should not throw if called before server started', () => {
    const gql = new GQLService({
      configDir: createTempFiles({
        '.gqlconfig': `
          {
            schema: {
              files: 'schema/*.gql',
            }
          }
        `,
        'schema/schema.gql': `
          type Query {
            viewer: Viewer
          }

          type Viewer {
            name: String
          }
        `,
      }),
      watch: false,
    });

    const run = () =>
      gql.autocomplete({
        sourcePath: 'schema/user.gql',
        ...code(`
        type User {
          viewer: Vi
          #---------^
        }
      `),
      });

    expect(run).not.toThrow();
    expect(run()).toEqual([]);
  });
});

const allHints = getHints();

describe('Schema', () => {
  describe('type', () => {
    it('show all types for Field', async () => {
      const hints = await autocomplete({
        sourcePath: 'schema/test.graphql',
        ...code(`
          type TestType {
            name:   #
            #-----^
          }
        `),
      });
      expect(hints).toEqual(allHints.OutputTypes);
    });

    it('show all types for Field even after typing few chars', async () => {
      const hints = await autocomplete({
        sourcePath: 'schema/test.graphql',
        ...code(`
          type TestType {
            name: S  #
            #------^
          }
        `),
      });
      expect(hints).toEqual(allHints.OutputTypes);
    });

    it('show all output types for Field if cursor on ! and of the form T!', async () => {
      const hints = await autocomplete({
        sourcePath: 'schema/test.graphql',
        ...code(`
          type TestType {
            name: S!   ##
            #-------^
          }
        `),
      });
      expect(hints).toEqual(allHints.OutputTypes);
    });

    it('show all output types for Field if cursor on ] and of type [T]', async () => {
      const hints = await autocomplete({
        sourcePath: 'schema/test.graphql',
        ...code(`
          type TestType {
            name: [S] #
            #-------^
          }
        `),
      });
      expect(hints).toEqual(allHints.OutputTypes);
    });

    it('show all output types for Field if cursor on ! and of the form [T!]', async () => {
      const hints = await autocomplete({
        sourcePath: 'schema/test.graphql',
        ...code(`
          type TestType {
            name: [S!]
            #-------^
          }
        `),
      });
      expect(hints).toEqual(allHints.OutputTypes);
    });

    it('dont show any types for Field if cursor after !', async () => {
      const hints = await autocomplete({
        sourcePath: 'schema/test.graphql',
        ...code(`
          type TestType {
            name: S!  #
            #--------^
          }
        `),
      });
      expect(hints).toEqual([]);
    });

    it('show all types for Field with args', async () => {
      const hints = await autocomplete({
        sourcePath: 'schema/test.graphql',
        ...code(`
          type TestType {
            players(first: String!):   #
                            #--------^
          }
        `),
      });
      expect(hints).toEqual(allHints.OutputTypes);
    });

    it('show all types for Field ListType', async () => {
      const hints = await autocomplete({
        sourcePath: 'schema/test.graphql',
        ...code(`
          type TestType {
            name: []
            #-----^ (inside bracket)
          }
        `),
      });
      expect(hints).toEqual(allHints.OutputTypes);
    });

    it('show only InputType for Field input args ', async () => {
      const hints = await autocomplete({
        sourcePath: 'schema/test.graphql',
        ...code(`
          type TestType {
            name(xyz: )
            #--------^
          }
        `),
      });
      expect(hints).toEqual(allHints.InputTypes);
    });

    it('show `implements` && `{` after typename', async () => {
      const hints = await autocomplete({
        sourcePath: 'schema/test.graphql',
        ...code(`
          type TestType   #
          #-------------^
        `),
      });
      expect(hints).toMatchSnapshot();
    });

    it('show all `GraphQLInterfaceType` types after `implements` keyword', async () => {
      const hints = await autocomplete({
        sourcePath: 'schema/test.graphql',
        ...code(`
          type TestType implements   # (Dont remove hash will avoid removing space after line)
          #------------------------^
        `),
      });
      expect(hints).toEqual(allHints.InterfaceTypes);
    });

    it('show all `GraphQLInterfaceType` types after multiple InterfaceTypes', async () => {
      const hints = await autocomplete({
        sourcePath: 'schema/test.graphql',
        ...code(`
          type TestType implements Node   #
                #-----------------------^
        `),
      });
      expect(hints).toEqual(allHints.InterfaceTypes);
    });
  });

  describe('input', () => {
    it('show only InputType for field', async () => {
      const hints = await autocomplete({
        sourcePath: 'schema/test.graphql',
        ...code(`
          input TestTypeInput {
            name:   #
            ------^
          }
        `),
      });
      expect(hints).toEqual(allHints.InputTypes);
    });

    it('show only InputType for field of ListType', async () => {
      const hints = await autocomplete({
        sourcePath: 'schema/test.graphql',
        ...code(`
          input TestType {
            name: []
            ------^ (inside bracket)
          }
        `),
      });
      expect(hints).toEqual(allHints.InputTypes);
    });
  });

  describe('union', () => {
    it('show only GrahQLObjectType for field', async () => {
      const hints = await autocomplete({
        sourcePath: 'schema/test.graphql',
        ...code(`
          union TestUnionType =   #
              #-----------------^
        `),
      });
      expect(hints).toEqual(allHints.ObjectTypes);
    });

    it('show all GrahQLObjectType after `|`', async () => {
      const hints = await autocomplete({
        sourcePath: 'schema/test.graphql',
        ...code(`
          union TestUnionType =  TestType |   #
                        #-------------------^
        `),
      });
      expect(hints).toEqual(allHints.ObjectTypes);
    });
  });
});

describe('Query', () => {
  describe('fragments', () => {
    it('on Type suggestions', async () => {
      const result = await autocomplete({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on   #
              ---------------^
          \`;
        `),
      });
      expect(result).toEqual(allHints.CompositeTypes);
    });

    it('on Type suggestions after few characters', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on Pl  #
              ----------------^
          \`;
        `),
      });
      expect(hints).toEqual(allHints.CompositeTypes);
    });

    it('fields suggestions: Simple', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on Player {
                    #
            #---^
            }
          \`;
        `),
      });

      expect(hints).toEqual(allHints.PlayerTypeFields);
    });

    it('alias fields', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on Player {
              some:    #
              ------^
            }
          \`;
        `),
      });

      expect(hints).toEqual(allHints.PlayerTypeFields);
    });

    it('[Relay] fields after interpolation', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on Player {
              id
              \${Component.getFragment('player')}
                    #
              #---^
            }
          \`;
        `),
      });

      expect(hints).toEqual(allHints.PlayerTypeFields);
    });

    it('fields args', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on Player {
              image()
              ------^
            }
          \`;
        `),
      });
      expect(hints).toEqual(allHints.PlayerTypeImageFieldArgs);
    });

    it('inline fragments on Type', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on Node {
              id
              ...on   #
              ------^
            }
          \`;
        `),
      });
      expect(hints).toEqual(allHints.TypesImplementsNode);
    });

    it('inline fragments on Type fields', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment test on Node {
              id
              ...on Player {
                    #
              #--^
              }
            }
          \`;
        `),
      });
      expect(hints).toEqual(allHints.PlayerTypeFields);
    });
  });

  describe('mutation', () => {
    it('fields', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay/test.js',
        ...code(`
          Relay.QL\`
            mutation { }
            ----------^
          \`;
        `),
      });
      expect(hints).toEqual(allHints.PossibleMutations);
    });

    it('field args', async () => {
      const hints = await autocomplete({
        sourcePath: 'query/test.graphql',
        ...code(`
          mutation {
            PlayerCreate()
            -------------^
          }
        `),
      });
      expect(hints).toMatchSnapshot();
    });

    it('variable definition types', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay-modern/test.js',
        ...code(`
          graphql\`
            mutation PlayerCreate($input: C )
                      #--------------------^
          \`
        `),
      });
      expect(hints).toEqual(allHints.InputTypes);
    });
  });

  describe('query', () => {
    it('fields', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay/test.js',
        ...code(`
          Relay.QL\`
            query Test { }
            ------------^
          \`;
        `),
      });
      expect(hints).toEqual(allHints.QueryFields);
    });
  });

  describe('subscription', () => {
    it('fields', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay/test.js',
        ...code(`
          Relay.QL\`
            subscription { }
            --------------^
          \`;
        `),
      });
      expect(hints).toEqual(allHints.PossibleSubscriptions);
    });

    it('field args', async () => {
      const hints = await autocomplete({
        sourcePath: 'query/test.graphql',
        ...code(`
          subscription {
            LikeStory()
            ----------^
          }
        `),
      });
      expect(hints).toMatchSnapshot();
    });
  });

  describe('values', () => {
    it('enum value suggesstion', async () => {
      const hints = await autocomplete({
        sourcePath: 'query/test.gql',
        ...code(`
          fragment test on NewPlayer {
            image(role: )
            #----------^
          }
        `),
      });
      expect(hints).toEqual(allHints.EnumRoleValues);
    });

    it('Object value keys', async () => {
      const hints = await autocomplete({
        sourcePath: 'query/test.graphql',
        ...code(`
          mutation {
            PlayerCreate(
              input: {
                id: "string",
                  #
            #---^
              }
            })
          }
        `),
      });
      expect(hints).toEqual(allHints.PlayerCreateInputFields);
    });
  });

  describe('directives', () => {
    it('directives', async () => {
      const hints = await autocomplete({
        sourcePath: 'query/test.gql',
        ...code(`
          fragment on Test {
            id @inc
          --------^
          }
        `),
      });
      expect(hints).toMatchSnapshot();
    });

    it('directives: include relay directives', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay/test.js',
        ...code(`
          Relay.QL\`
            fragment on Test @relay {
                    #----------^
            }
          \`;
        `),
      });
      expect(hints).toMatchSnapshot();
    });

    it('args', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay/test.js',
        ...code(`
          Relay.QL\`
            fragment on Test @relay() {
                --------------------^
              id
            }
          \`;
        `),
      });
      expect(hints).toMatchSnapshot();
    });

    it('args: value', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay/test.js',
        ...code(`
          Relay.QL\`
            fragment on Test @relay(pattern:  ) {
                                    ---------^
              id
            }
          \`;
        `),
      });
      expect(hints).toMatchSnapshot();
    });
  });

  describe('fragmentSpread', () => {
    test('show fragments defined in same document', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay-modern/test.js',
        ...code(`
            const a = graphql\`
              fragment Test on Player {
                id
                ...
              #----^
              }

              fragment ProfileA on Player {
                id
              }
            \`
          `),
      });
      expect(hints).toMatchSnapshot();
    });

    test('show fragments defined in same file', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay-modern/test.js',
        ...code(`
          const a = graphql\`
            fragment Test on Player {
              id
              ...
            #----^
            }
          \`

          const b = graphql\`
            fragment ProfileB on Player {
              id
            }
          \`
        `),
      });
      expect(hints).toMatchSnapshot();
    });

    test('show fragments defined in other files', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay-modern/test.js',
        ...code(`
          const a = graphql\`
            fragment Test on Player {
              id
              ...
            #----^
            }
          \`
        `),
        otherFiles: {
          'relay-modern/test2.js': `
            const c = graphql\`
              fragment ProfileC on Player {
                id
              }
            \`
          `,
        },
      });
      expect(hints).toMatchSnapshot();
    });
  });

  describe('meta field __typename', () => {
    it('interface type', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment on Node {
              id
              __type
            #-------^
            }
          \`
        `),
      });
      expect(hints).toContainEqual(allHints.TypeNameMetaField);
    });

    it('union type', async () => {
      const hints = await autocomplete({
        sourcePath: 'relay/test.js',
        ...code(`
          const a = Relay.QL\`
            fragment on Entity {
              __type
            #-------^
            }
          \`
        `),
      });
      expect(hints).toContainEqual(allHints.TypeNameMetaField);
    });
  });

  describe('meta field __schema', () => {
    test('Query', async () => {
      const hints = await autocomplete({
        sourcePath: 'query/test.graphql',
        ...code(`
          query {
            __
        #---^
          }
        `),
      });
      expect(hints).toContainEqual(allHints.SchemaMetaField);
    });
  });

  describe('meta field __type', () => {
    test('Query', async () => {
      const hints = await autocomplete({
        sourcePath: 'query/test.graphql',
        ...code(`
          query {
            __
        #---^
          }
        `),
      });
      expect(hints).toContainEqual(allHints.TypeMetaField);
    });
  });

  describe('[Bug] should work if position is first character', () => {
    it('whitespace', async () => {
      const hints = await autocomplete({
        sourcePath: 'query/test.graphql',
        sourceText: ' ',
        position: { line: 1, column: 1 },
      });
      expect(hints).toEqual(allHints.DocumentLevel);
    });

    it('mutation keyword', async () => {
      const hints = await autocomplete({
        sourcePath: 'query/test.graphql',
        sourceText: 'mutation',
        position: { line: 1, column: 1 },
      });
      expect(hints).toMatchSnapshot();
    });

    it('fragment keyword', async () => {
      const hints = await autocomplete({
        sourcePath: 'query/test.graphql',
        sourceText: 'fragment',
        position: { line: 1, column: 1 },
      });
      expect(hints).toEqual([{ text: 'fragment' }]);
    });
  });

  test('[Bug] should work with empty sourceText', async () => {
    const hints = await autocomplete({
      sourcePath: 'query/test.graphql',
      sourceText: '',
      position: { line: 1, column: 1 },
    });
    expect(hints).toEqual(allHints.DocumentLevel);
  });
});

async function autocomplete(options: {
  sourcePath: string,
  sourceText: string,
  position: any,
  otherFiles?: Object,
}) {
  const rootDir = createTempFiles({
    '.gqlconfig': `
        {
          schema: {
            files: ['schema/*.gql', 'schema/**/*.graphql'],
            graphQLOptions: {
              commentDescriptions: true
            },
          },
          query: {
            files: [
              {
                match: ['query/**/*.graphql', 'query/**/*.gql'],
              },
              {
                match: 'relay/**/*.js',
                presets: ['relay'],
              },
              {
                match: 'relay-modern/**/*.js',
                presets: ['relay-modern'],
              },
              {
                match: 'apollo/**/*.js',
                presets: ['apollo'],
              },
            ]
          }
        }
      `,
    ...getSchemaFiles(),
    ...options.otherFiles,
  });

  const gql = new GQLService({
    configDir: rootDir,
    watch: false,
  });
  gql.onError(err => {
    throw err;
  });

  await gql.start();
  const result = gql
    .autocomplete({
      sourcePath: path.join(rootDir, options.sourcePath),
      sourceText: options.sourceText,
      position: options.position,
    })
    .sort(sortHints);
  await gql.stop();
  return result;
}
