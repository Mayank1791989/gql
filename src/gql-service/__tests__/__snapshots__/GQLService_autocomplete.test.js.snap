// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query [Bug] should work if position is first character mutation keyword 1`] = `
Array [
  Object {
    "description": "Mutation contains all allowed mutations",
    "text": "mutation",
    "type": "Mutation",
  },
]
`;

exports[`Query directives args 1`] = `
Array [
  Object {
    "description": "Marks this fragment spread as being deferrable such that it loads after other portions of the view.",
    "text": "deferrable",
    "type": "Boolean",
  },
  Object {
    "description": "Marks a connection field as containing nodes without 'id' fields. This is used to silence the warning when diffing connections.",
    "text": "isConnectionWithoutNodeID",
    "type": "Boolean",
  },
  Object {
    "description": "Marks a fragment spread which should be unmasked if provided false",
    "text": "mask",
    "type": "Boolean",
  },
  Object {
    "description": "Marks a fragment as intended for pattern matching (as opposed to fetching). Used in Classic only.",
    "text": "pattern",
    "type": "Boolean",
  },
  Object {
    "description": "Marks a fragment as being backed by a GraphQLList.",
    "text": "plural",
    "type": "Boolean",
  },
  Object {
    "description": "Selectively pass variables down into a fragment. Only used in Classic.",
    "text": "variables",
    "type": "[String!]",
  },
]
`;

exports[`Query directives args: value 1`] = `
Array [
  Object {
    "description": "Not true.",
    "text": "false",
    "type": "Boolean",
  },
  Object {
    "description": "Not false.",
    "text": "true",
    "type": "Boolean",
  },
]
`;

exports[`Query directives directives 1`] = `
Array [
  Object {
    "description": "some custom directive",
    "text": "customDirective",
    "type": "Directive",
  },
  Object {
    "description": "Directs the executor to include this field or fragment only when the \`if\` argument is true.",
    "text": "include",
    "type": "Directive",
  },
  Object {
    "description": "Directs the executor to skip this field or fragment when the \`if\` argument is true.",
    "text": "skip",
    "type": "Directive",
  },
]
`;

exports[`Query directives directives: include relay directives 1`] = `
Array [
  Object {
    "description": "some custom directive",
    "text": "customDirective",
    "type": "Directive",
  },
  Object {
    "description": "",
    "text": "relay",
    "type": "Directive",
  },
]
`;

exports[`Query fragmentSpread show fragments defined in other files 1`] = `
Array [
  Object {
    "description": "fragment ProfileC on Player",
    "text": "ProfileC",
    "type": "Player",
  },
]
`;

exports[`Query fragmentSpread show fragments defined in same document 1`] = `
Array [
  Object {
    "description": "fragment ProfileA on Player",
    "text": "ProfileA",
    "type": "Player",
  },
]
`;

exports[`Query fragmentSpread show fragments defined in same file 1`] = `
Array [
  Object {
    "description": "fragment ProfileB on Player",
    "text": "ProfileB",
    "type": "Player",
  },
]
`;

exports[`Query mutation field args 1`] = `
Array [
  Object {
    "description": "",
    "text": "input",
    "type": "PlayerCreateInput!",
  },
]
`;

exports[`Query subscription field args 1`] = `
Array [
  Object {
    "description": "",
    "text": "input",
    "type": "LikeStorySubscriptionInput",
  },
]
`;

exports[`Schema type show \`implements\` && \`{\` after typename 1`] = `
Array [
  Object {
    "text": "{",
  },
  Object {
    "text": "implements",
    "type": "Implements",
  },
]
`;

exports[`Schema: autocomplete works in schema files 1`] = `
Array [
  Object {
    "description": "A Directive can be adjacent to many parts of the GraphQL language, a __DirectiveLocation describes one such possible adjacencies.",
    "text": "__DirectiveLocation",
    "type": "Enum",
  },
  Object {
    "description": "An enum describing what kind of type a given \`__Type\` is.",
    "text": "__TypeKind",
    "type": "Enum",
  },
  Object {
    "description": undefined,
    "text": "Query",
    "type": "Object",
  },
  Object {
    "description": undefined,
    "text": "Viewer",
    "type": "Object",
  },
  Object {
    "description": "A Directive provides a way to describe alternate runtime execution and type validation behavior in a GraphQL document.

In some cases, you need to provide options to alter GraphQL's execution behavior in ways field arguments will not suffice, such as conditionally including or skipping a field. Directives provide this by describing additional information to the executor.",
    "text": "__Directive",
    "type": "Object",
  },
  Object {
    "description": "One possible value for a given Enum. Enum values are unique values, not a placeholder for a string or numeric value. However an Enum value is returned in a JSON response as a string.",
    "text": "__EnumValue",
    "type": "Object",
  },
  Object {
    "description": "Object and Interface types are described by a list of Fields, each of which has a name, potentially a list of arguments, and a return type.",
    "text": "__Field",
    "type": "Object",
  },
  Object {
    "description": "Arguments provided to Fields or Directives and the input fields of an InputObject are represented as Input Values which describe their type and optionally a default value.",
    "text": "__InputValue",
    "type": "Object",
  },
  Object {
    "description": "A GraphQL Schema defines the capabilities of a GraphQL server. It exposes all available types and directives on the server, as well as the entry points for query, mutation, and subscription operations.",
    "text": "__Schema",
    "type": "Object",
  },
  Object {
    "description": "The fundamental unit of any GraphQL Schema is the type. There are many kinds of types in GraphQL as represented by the \`__TypeKind\` enum.

Depending on the kind of a type, certain fields describe information about that type. Scalar types provide no information beyond a name and description, while Enum types provide their values. Object and Interface types provide the fields they describe. Abstract types, Union and Interface, provide the Object types possible at runtime. List and NonNull types compose other types.",
    "text": "__Type",
    "type": "Object",
  },
  Object {
    "description": "The \`Boolean\` scalar type represents \`true\` or \`false\`.",
    "text": "Boolean",
    "type": "Scalar",
  },
  Object {
    "description": "The \`Float\` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ",
    "text": "Float",
    "type": "Scalar",
  },
  Object {
    "description": "The \`ID\` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as \`\\"4\\"\`) or integer (such as \`4\`) input value will be accepted as an ID.",
    "text": "ID",
    "type": "Scalar",
  },
  Object {
    "description": "The \`Int\` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ",
    "text": "Int",
    "type": "Scalar",
  },
  Object {
    "description": "The \`String\` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.",
    "text": "String",
    "type": "Scalar",
  },
]
`;
